File: Blockchain/exp01/02_PoW/go/main/pow.go
package main

import (
	"bytes"
	"crypto/sha256"
	"encoding/binary"
	"fmt"
	"log"
	"math"
	"math/big"
	"strconv"
	// "time"
)

var (
	maxNonce = math.MaxInt64
)

const targetBits = 8

func IntToHex(num int64) []byte {
	buff := new(bytes.Buffer)
	err := binary.Write(buff, binary.BigEndian, num)
	if err != nil {
		log.Panic(err)
	}

	return buff.Bytes()
}

type ProofOfWork struct {
	block  *Block
	target *big.Int
}

func NewProofOfWork(b *Block) *ProofOfWork {
	target := big.NewInt(1)
	target.Lsh(target, uint(256-targetBits))

	pow := &ProofOfWork{b, target}
	return pow
}

func (pow *ProofOfWork) prepareData(nonce int) []byte {
	data := bytes.Join(
		[][]byte{
			pow.block.PrevBlockHash,
			pow.block.Data,
			IntToHex(pow.block.TimeStamp),
			IntToHex(int64(targetBits)),
			IntToHex(int64(nonce)),
		},
		[]byte{},
	)

	return data
}

func (pow *ProofOfWork) Validate() bool {
	var hashInt big.Int

	data := pow.prepareData(pow.block.Nonce)
	hash := sha256.Sum256(data)
	hashInt.SetBytes(hash[:])

	isValid := hashInt.Cmp(pow.target) == -1

	return isValid
}

func (pow *ProofOfWork) Run() (int, []byte) {
	var hashInt big.Int
	var hash [32]byte
	nonce := 0

	fmt.Printf("Mining the block containing \"%s\"\n", pow.block.Data)

	for nonce < maxNonce {
		data := pow.prepareData(nonce)

		hash = sha256.Sum256(data)
		hashInt.SetBytes(hash[:])

		if hashInt.Cmp(pow.target) == -1 {
			fmt.Printf("%x\t\n", data)
			// fmt.Printf("%s:\t\n", hashInt)
			fmt.Printf("%x\t\n", hash)
			// fmt.Printf("\r%x", hash)
			break
		} else {
			nonce++
		}
	}
	fmt.Printf("\n\n")

	return nonce, hash[:]
}

type Block struct {
	/* block headers */
	TimeStamp     int64  // 当前时间戳，即区块创建的时间
	PrevBlockHash []byte // 前一个区块的Hash
	Hash          []byte // 当前区块的Hash
	/* body */
	Data  []byte // 区块实际存储的信息, 即比特币中的 Transaction
	Nonce int
}

func NewBlock(data string, prevBlockHash []byte) *Block {
	block := &Block{
		// TimeStamp:     time.Now().Unix(),
		TimeStamp:     0,
		PrevBlockHash: prevBlockHash,
		Hash:          []byte{},
		Data:          []byte(data),
		Nonce:         0,
	}
	pow := NewProofOfWork(block)
	nonce, hash := pow.Run()

	block.Hash = hash[:]
	block.Nonce = nonce

	return block
}

type BlockChain struct {
	blocks []*Block
}

//
func (bc *BlockChain) AddBlock(data string) {
	prevBlock := bc.blocks[len(bc.blocks)-1]
	newBlock := NewBlock(data, prevBlock.Hash)
	bc.blocks = append(bc.blocks, newBlock)
}

func NewGenesisBlock() *Block {
	return NewBlock("Genesis Block", []byte{})
}

func NewBlockChain() *BlockChain {
	return &BlockChain{[]*Block{NewGenesisBlock()}}
}

func main() {
	bc := NewBlockChain()

	bc.AddBlock("send 1 BTC to Ivan")
	bc.AddBlock("send 2 more BTC to Ivan")

	for _, block := range bc.blocks {
		fmt.Printf("Prev hash: %x\n", block.PrevBlockHash)
		fmt.Printf("Data: %s\n", block.Data)
		fmt.Printf("Hash: %x\n", block.Hash)
		pow := NewProofOfWork(block)
		fmt.Printf("PoW: %s\n", strconv.FormatBool(pow.Validate()))
		fmt.Println()
	}
}

File: Blockchain/exp01/02_PoW/cpp/pow.cpp
#include <vector>
#include <iostream>
#include <ctime>	// http://www.cplusplus.com/reference/ctime/time/
#include <iomanip>
#include <sstream>
#include <openssl/sha.h>
#include <climits>			// http://www.cplusplus.com/reference/climits/
#define u8 unsigned char
#define u32 unsigned int
#define u64 unsigned long long
#define maxNonce ULLONG_MAX
using namespace std;

// https://stackoverflow.com/questions/5100718/integer-to-hex-string-in-c
string int2hex(u64 num){
	stringstream ss;
	ss<<setw(16)<<setfill('0')<<hex<<num;
	return ss.str();
}

// TODO
// string s = "000000000000000100000000000000000000000000000000";
// strHexTargetHash = strBin2StrHex(s, 32);
// strHexTargetHash = "0100000000000000000000000000000000000000000000000000000000000000";
// 0000 0000 (b) -> 00
// 0000 0001 (b) -> 01
// 0000 0010 (b) -> 02
string strBin2StrHex(string num, size_t size){
	stringstream ss;
	for( size_t i = 0; i < size; i += 8 ){
		string strPiece = num.substr(i, 8);
		u32 intPiece = (u32)strtoul(strPiece.c_str(), NULL, 2);
		ss<<hex<<setw(2)<<setfill('0')<<intPiece;
	}
	return ss.str();
}

// https://stackoverflow.com/questions/2262386/generate-sha256-with-openssl-and-c
// https://stackoverflow.com/questions/13784434/how-to-use-openssls-sha256-functions
// https://www.xorbin.com/tools/sha256-hash-calculator
string getsha256(string str){
	unsigned char hash[SHA256_DIGEST_LENGTH];
	SHA256_CTX sha256;
	SHA256_Init(&sha256);
	SHA256_Update(&sha256, str.c_str(), str.size());
	SHA256_Final(hash, &sha256);
	stringstream ss;
	for(int i = 0; i < SHA256_DIGEST_LENGTH; i++){
		ss << hex << setw(2) << setfill('0') << (int)hash[i];
	}
	return ss.str();
}

// 将字节数组，转换成字符串
// 举例
// 41 -> 0x41 -> 65 -> A
// 字节数组用字符串表示，是为了方便打印查看 hash 
string byteArr2str(string strByteArray){
	stringstream ss;
	for( int i = 0; i < strByteArray.size(); i += 2 ){
		string strhex = strByteArray.substr(i, 2);
		// https://stackoverflow.com/questions/17261798/converting-a-hex-string-to-a-byte-array
		// http://www.cplusplus.com/reference/cstdlib/strtoul/?kw=strtoul
		ss<<(u8)strtoul(strhex.c_str(), NULL, 16);
	}
	string s = ss.str();
	return s;
}

const size_t targetBits = 8;

struct Block{
	u64 nonce;
	u64 timeStamp;
	string msg;
	string prevHash;		// byte array 的字符串形式
	string currHash;		// byte array 的字符串形式
	string strHexTargetHash;
	Block * next;

	Block(string _msg, string pH){
		msg = _msg;
		prevHash = pH;

		// 测试一致性
		// 对 go 做同样处理，使两者的 timestamp 都为 0，控制变量
		// 最后的结果应该是一样的
		timeStamp = 0;
		// 实际应用时，取消下面 3 行的注释
		// time_t timer;
		// time(&timer);
		// timeStamp = (u64)timer;

		if( targetBits > 0 && targetBits < 256 ){
			// 如果 targetBit = 4, 则 target <<= 252
			// 得到 target = 000100000...000 (binary)
			// 若   hash   < 000100000...000 (binary)
			// 则   hash  == 0000xxxxx...xxx
			// 即 hash 前4位为 0, 即为有效 hash
			strHexTargetHash = string(255, '0');
			strHexTargetHash.insert(targetBits - 1, "1");
			strHexTargetHash = strBin2StrHex(strHexTargetHash, 256);
		}

		mining();

		next = NULL;
	}

	void mining(){
		cout<<"Mining the block containing \""<<msg<<"\"\n";
		for( u64 tmpNonce = 0; tmpNonce < maxNonce; tmpNonce++ ){
			string data = prepareData( tmpNonce );
			string tmpHash = getsha256(data);

			// https://wizardforcel.gitbooks.io/golang-stdlib-ref/content/81.html#Int.Cmp
			if( tmpHash < strHexTargetHash ){
				// 小于，则前 targetBits 位都是0, 即符合要求
				cout<<tmpHash<<"\n\n";
				currHash = tmpHash;
				nonce = tmpNonce;
				break;
			}
		}
	}

	string prepareData(u64 tmpNonce){
		// prototype 中, timestamp 是 strconv.FormatInt(b.TimeStamp, 10)
		// 这里变成了 hex
		string data = byteArr2str(prevHash) + msg +
					byteArr2str(int2hex(timeStamp)) +
					byteArr2str(int2hex(targetBits)) +
					byteArr2str(int2hex(tmpNonce));
		return data;
	}

	bool validate(){
		string data = prepareData(nonce);
		string tmpHash = getsha256(data);
		return tmpHash < strHexTargetHash;
	}
};

struct BlockChain{
	Block *head, *tail;
	int size;

	BlockChain(){
		head = new Block("Genesis Block", "");
		tail = head;
		size = 1;
	}

	int addBlock(string msg){
		tail->next = new Block(msg, tail->currHash);
		tail = tail->next;
		size++;
	}
};

int main(int argc, char *argv[]){
	BlockChain * bc = new BlockChain();
	bc->addBlock("send 1 BTC to Ivan");
	bc->addBlock("send 2 more BTC to Ivan");

	cout<<"\n\n\n";
	Block * p = bc->head;
	for( size_t i = 0; i < bc->size; i++ ){
		cout<<"prev hash: "<<p->prevHash<<'\n';
		cout<<"curr msg: "<<p->msg<<'\n';
		cout<<"curr Hash: "<<p->currHash<<'\n';
		cout<<"PoW: "<<p->validate()<<"\n\n\n";
		p = p->next;
	}

	return 0;
}

// TODO
// 3. 数据库(或者自己写保存文件)
// 4. 命令行
// 5. Python
